invoice/src/main/resources/org/killbill/billing/invoice/dao/InvoiceSqlDao.sql.stg

```diff
@@ -95,61 +89,78 @@ getInvoiceByAccountRecordIdAfter() ::= <<
    <defaultOrderBy("")>
 >>
 
-searchByAccountOrInvoiceId(ordering) ::= <<
-(select
-<allTableFields("t.")>
-from <tableName()> t
-where (<idField("t.")> = :searchKey)
-<andCheckSoftDeletionWithComma("t.")>
-<AND_CHECK_TENANT("t.")>)
-UNION
-(select
-<allTableFields("t.")>
-from <tableName()> t
-where (t.account_id = :searchKey)
-<andCheckSoftDeletionWithComma("t.")>
-<AND_CHECK_TENANT("t.")>)
-order by <recordIdField("")> <ordering>
-limit :rowCount offset :offset
-;
->>
-
-getSearchByAccountOrInvoiceIdCount() ::= <<
-select count(*) from
-((select
-<allTableFields("t.")>
-from <tableName()> t
-where (<idField("t.")> = :searchKey)
-<andCheckSoftDeletionWithComma("t.")>
-<AND_CHECK_TENANT("t.")>)
-UNION
-(select
-<allTableFields("t.")>
-from <tableName()> t
-where (t.account_id = :searchKey)
-<andCheckSoftDeletionWithComma("t.")>
-<AND_CHECK_TENANT("t.")>)) as count
-;
+invoiceBalanceQuery() ::= <<
+SELECT
+      x.invoice_id
+    , x.tenant_record_id
+    , SUM(x.amount) as balance
+FROM (
+    SELECT
+          i.id as invoice_id
+        , i.tenant_record_id
+        , 0 as amount
+    FROM invoices i
+    WHERE
+        <CHECK_TENANT("i.")>
+
+    UNION ALL
+
+    SELECT
+          ii.invoice_id
+        , ii.tenant_record_id
+        , SUM(ii.amount) as amount
+    FROM invoice_items ii
+    WHERE
+        ii.type in ('TAX', 'EXTERNAL_CHARGE', 'FIXED', 'USAGE', 'RECURRING', 'ITEM_ADJ', 'REPAIR_ADJ', 'PARENT_SUMMARY', 'CBA_ADJ', 'CREDIT_ADJ')
+        <AND_CHECK_TENANT("ii.")>
+    GROUP BY
+          ii.invoice_id
+        , ii.tenant_record_id
+
+    UNION ALL
+
+    SELECT
+          invoice_id
+        , tenant_record_id
+        , -payments_and_refunds.amount_paid - payments_and_refunds.amount_refunded as amount
+    FROM (
+        SELECT
+              ip.invoice_id
+            , ip.tenant_record_id
+            , SUM(CASE WHEN ip.type = 'ATTEMPT' THEN ip.amount ELSE 0 END) AS amount_paid
+            , SUM(CASE WHEN ip.type IN ('REFUND', 'CHARGED_BACK') THEN ip.amount ELSE 0 END) AS amount_refunded
+        FROM
+            invoice_payments ip
+        WHERE
+            ip.status = 'SUCCESS'
+            <AND_CHECK_TENANT("ip.")>
+        GROUP BY
+              ip.invoice_id
+            , ip.tenant_record_id
+    ) AS payments_and_refunds
+) x
+GROUP BY
+      x.invoice_id
+    , x.tenant_record_id
 >>
 
-searchByCurrency(ordering) ::= <<
+searchInvoicesByBalance(ordering, comparisonOperator) ::= <<
```

Comment:
The result(s) we return will match the criteria, but since we just return the columns (fileds), the exact balance will not be returned, is this correct - i.e. balance computation is just used for search criteria and not to populate the objects being returned?

Reply:
> balance computation is just used for search criteria and not to populate the objects being returned

Correct.

> since we just return the columns (fileds)

That's a good point, I need to check if the objects are re-hydrated at higher levels of the stack or if I should select more columns here.

Reply:
I am just wondering if we could use the `amount` computed from the underlying `invoiceBalanceQuery()` so we don;'t have to rehydrate the invoices. as it is very expensive. However, not returning the balance is an issue as well.

Reply:
I am just wondering if we could use the `amount` computed from the underlying `invoiceBalanceQuery()` so we don't have to rehydrate the invoices. as it is very expensive. However, not returning the balance is an issue as well.

Reply:
We could:

* Modify `InvoiceModelDao` to have a `balance` field. This is set only in this new codepath, when computing the balance in SQL (the `balance` column from the SQL query should automatically populate it if the getter and setter for balance are added in `InvoiceModelDao`.
* Modify `DefaultInvoice` to have a private `balance` field. If set during construction of the object (pulled from `InvoiceModelDao`), `getBalance()` returns it, otherwise, it goes through the computation like today.

@reshmabidikar Could you give this a try? When doing searches, the invoices will still be shallow, but when `searchInvoicesByBalance` is called, the balance should be returned. If this work, we could also look at updating the other search queries to do the same?